### PR TITLE
Add support for "update" command to fwupdtool

### PR DESCRIPTION
**Pull request by superm1**
Type of pull request:
- [x] Feature

This is intended to be used with local repositories/metadata sources that are already configured.

This was tested using the changes here: https://github.com/hughsie/lvfs-website/pull/233

### CODE REVIEW

File: src/fu-tool.c [hunk 1]
```diff
@@ -75,6 +76,9 @@ fu_util_start_engine (FuUtilPrivate *priv, GError **error)
 	g_autoptr(GVariant) val = NULL;
 	g_autoptr(GError) error_local = NULL;
 
+	if (priv->engine_loaded)
```
Comment (hughsie) [inline src/fu-tool.c]: Maybe just create a `gboolean fu_engine_is_loaded()`?

File: src/fu-tool.c [hunk 2]
```diff
@@ -752,6 +760,83 @@ fu_util_install (FuUtilPrivate *priv, gchar **values, GError **error)
 	/* success */
 	return TRUE;
 }
+static gboolean
```
Comment (hughsie) [inline src/fu-tool.c]: space here pls

File: src/fu-tool.c [hunk 3]
```diff
@@ -752,6 +760,83 @@ fu_util_install (FuUtilPrivate *priv, gchar **values, GError **error)
...
+		g_autoptr(GError) error_local = NULL;
+		device_id = fu_device_get_id (dev);
+
+		if (!fu_util_is_interesting_device (dev))
+			continue;
+		/* only show stuff that has metadata available */
+		if (!fwupd_device_has_flag (dev, FWUPD_DEVICE_FLAG_SUPPORTED))
+			continue;
+
+		rels = fu_engine_get_upgrades (priv->engine, device_id, &error_local);
```
Comment (hughsie) [inline src/fu-tool.c]: you could move the `device_id = fu_device_get_id (dev);` here

File: src/fu-tool.c [hunk 4]
```diff
@@ -752,6 +760,83 @@ fu_util_install (FuUtilPrivate *priv, gchar **values, GError **error)
...
+		}
+
+		rel = g_ptr_array_index (rels, 0);
+		uri_tmp = fwupd_release_get_uri (rel);
+		remote_id = fwupd_release_get_remote_id (rel);
+		if (remote_id != NULL) {
+			FwupdRemote *remote;
+			g_autoptr(GPtrArray) remotes = NULL;
+			g_autofree gchar *fn = NULL;
+			gchar* install_args[2] = {NULL};
```
Comment (hughsie) [inline src/fu-tool.c]: `gchar *install_args[2] = { NULL };` pls

File: src/fu-tool.c [hunk 5]
```diff
@@ -752,6 +760,83 @@ fu_util_install (FuUtilPrivate *priv, gchar **values, GError **error)
...
+
+			remotes = fu_engine_get_remotes (priv->engine, error);
+			if (remotes == NULL)
+				return FALSE;
+
+			for (guint j = 0; j < remotes->len; j++) {
+				remote = g_ptr_array_index (remotes, j);
+				if (g_strcmp0 (remote_id, fwupd_remote_get_id (remote)) == 0)
+					break;
+				if (j+1 == remotes->len)
```
Comment (hughsie) [inline src/fu-tool.c]: space around the `+`

Comment (hughsie) [reply src/fu-tool.c]: Actually, this is pretty mad. Just create `fu_engine_get_remote_by_id()`

File: src/fu-tool.c [hunk 6]
```diff
@@ -752,6 +760,83 @@ fu_util_install (FuUtilPrivate *priv, gchar **values, GError **error)
...
+				g_printerr ("Couldn't find remote %s\n", remote_id);
+				continue;
+			}
+
+			/* local remotes have the firmware already */
+			if (fwupd_remote_get_kind (remote) == FWUPD_REMOTE_KIND_LOCAL) {
+				const gchar *fn_cache = fwupd_remote_get_filename_cache (remote);
+				g_autofree gchar *path = g_path_get_dirname (fn_cache);
+				fn = g_build_filename (path, uri_tmp, NULL);
+				install_args[0] = fn;
```
Comment (hughsie) [inline src/fu-tool.c]: bit eww. Maybe declare argv like `g_auto(GStrv) argv = g_new0 (gchar *, 2)` and then you can do `argv[0] = g_build_filename (path, uri_tmp, NULL);`

File: src/fu-tool.c [hunk 7]
```diff
@@ -752,6 +760,83 @@ fu_util_install (FuUtilPrivate *priv, gchar **values, GError **error)
...
+fu_util_update (FuUtilPrivate *priv, gchar **values, GError **error)
+{
+	g_autoptr(GPtrArray) devices = NULL;
+
+	/* load engine */
+	if (!fu_util_start_engine (priv, error))
+		return FALSE;
+
+	devices = fu_engine_get_devices (priv->engine, error);
+
```
Comment (hughsie) [inline src/fu-tool.c]: no need for space here

File: src/fu-tool.c (hunk 6 above)
Comment (superm1) [reply src/fu-tool.c]: That's much cleaner, thanks.

File: src/fu-tool.c (hunk 1 above)
Comment (superm1) [reply src/fu-tool.c]: Added the attribute to the engine instead, and let the engine have load called multiple times.

File: src/fu-tool.c [hunk 8]
```diff
@@ -753,6 +753,80 @@ fu_util_install (FuUtilPrivate *priv, gchar **values, GError **error)
...
+		rel = g_ptr_array_index (rels, 0);
+		uri_tmp = fwupd_release_get_uri (rel);
+		remote_id = fwupd_release_get_remote_id (rel);
+		if (remote_id != NULL) {
+			FwupdRemote *remote;
+			g_autoptr(GPtrArray) remotes = NULL;
+			g_auto(GStrv) argv = g_new0 (gchar *, 2);
+
+			remotes = fu_engine_get_remotes (priv->engine, error);
```
Comment (hughsie) [inline src/fu-tool.c]: I still think you want `fu_engine_get_remote_by_id()` :)

Comment (superm1) [reply src/fu-tool.c]: I just saw that earlier comment now.  I'll fixup.

**Pull request merged**